doc/specification/language-proposal.md

```diff
@@ -0,0 +1,1518 @@
+>
+> ***exponent-op*** ::= `^`
+>
+> ***preferential-union*** ::= `&` | U+2294 `⊔` SQUARE CAP
+>
+> ***compare-equal*** ::= `==`
+>
+> ***compare-not-equal*** ::= `!=` | U+2260 `≠` NOT EQUAL TO
+>
+> ***compare-less*** ::= `<`
+>
+> ***compare-less-equal*** ::= `<=` | U+2264 `≤` LESS-THAN OR EQUAL TO
+>
+> ***compare-greater*** ::= `>`
+>
+> ***compare-greater-equal*** ::= `>=` | U+2265 `≥` GREATER-THAN OR EQUAL TO
+>
+> ***left-arrow*** ::= `<-` | U+2190 `←` LEFTWARDS ARROW
+>
+> ***right-arrow*** ::= `->` | U+2192 `→` RIGHTWARDS ARROW
+>
+> ***right-left-arrow*** ::= `<->` | U+21C4 `⇄` RIGHT ARROW OVER LEFT ARROW
+>
+> ***empty-set*** ::= `\0` | U+2205 `∅` EMPTY SET
+>
+> ***square-root*** ::= U+221A `√` SQUARE ROOT
+>
+> ***assign-equal*** ::= `=`
+>
+> ***semicolon*** ::= `;`
+>
+> ***left-paren*** ::= `(`
+>
+> ***right-paren*** ::= `)`
+>
+> ***left-brace*** ::= `{`
+>
+> ***right-brace*** ::= `}`
+>
+> ***colon*** ::= `:`
+>
+> ***bar*** ::= U+007C `|` VERTICAL LINE
+>
+> ***comma*** ::= `,`
+>
+> ***period*** ::= `.`
+
+#### Token value
+
+Punctuation tokens have no value.
+
+
+# Arblang syntax and semantics
+<a id="Arblang-syntax-and-semantics"/>
+
+The syntax definitions below are defined in terms of the tokens defined in the lexical grammar, with the following conventions:
+
+* A ***symbol*** token can represent an identifier or a keyword; keywords in the syntax descriptions are written in `monospace`, and should be interpreted as ***symbol*** tokens with the corresponding value.
+
+* Where ***whitespace*** is present in a syntax rule, it is required, but additional ***whitespace*** tokens are permitted anywhere in the (tokenized) Arblang source, where they are ignored.
+
+* For readability in syntax rules, the text representation of a punctuation token, styled monospace, may stand in for the punctuation token itself. So, for example, `=` might stand for the token ***assign-equal***.
+
+An Arblang source document comprises a series of module and interface definitions.
+
+> _source_ ::= ( _module-defn_ | _interface-defn_ )+
+
+Module and interface definitions are described below.
+
+## Identifiers and scope
+<a id="identifiers-and-scope"/>
+
+Identifiers are represented by ***symbol*** tokens and depending on context, may refer to:
+1. a module,
+2. an imported module,
+3. a constant,
+4. a parameter,
+5. a function,
+6. an external binding,
+7. a type alias,
+8. a regime,
+9. a record field,
+10. a function argument,
+11. or a local value binding.
+
+### Scopes
+<a id="scopes"/>
+
+The association of an identifier with its referent is called a _binding_, and that binding is valid only within a _scope_. With one exception for regime names, an identifier's scope does not precede the point where that identifier is introduced by some declaration or binding.
+
+**Global scope.**
+A module definition introduces the module name into the following global scope. The environment may also introduce module names defined in other Arblang sources.
```

Comment:
The 'environment' wasn't ever defined, though it should be. It's necessarily handwavy though: it's meant to describe the extra bits and pieces relevant to the code but external to it, that are supplied by an interpreter or translator. In particular, it will include a possible set of modules available for import, but how that set is determined is external to the Arblang source.